File: src/auto_bdsp_rng/gen8_static/generator.py
from __future__ import annotations

from dataclasses import dataclass

from auto_bdsp_rng.gen8_static.models import (
    GENDER_FEMALE,
    GENDER_GENDERLESS,
    GENDER_MALE,
    GENDER_RATIO_FEMALE_ONLY,
    GENDER_RATIO_GENDERLESS,
    GENDER_RATIO_MALE_ONLY,
    Lead,
    Profile8,
    Shiny,
    State8,
    StateFilter,
    StaticTemplate8,
    get_shiny,
    hidden_power,
    is_shiny,
    validate_seed64,
)
from auto_bdsp_rng.rng_core.generators import BDSPXorshift, RNGList, XoroshiroBDSP
from auto_bdsp_rng.rng_core.seed import SeedPair64, U32_MAX

try:
    from auto_bdsp_rng.rng_core._native import generate_static as _native_generate
    _HAS_NATIVE = True
except ImportError:
    _HAS_NATIVE = False


def _gen_static_ec(rng: BDSPXorshift) -> int:
    return ((rng.next() % U32_MAX) + 0x80000000) & U32_MAX


def _is_synchronize(lead: Lead | int) -> bool:
    value = int(lead)
    return int(Lead.SYNCHRONIZE_START) <= value <= int(Lead.SYNCHRONIZE_END)


def _force_shiny(pid: int, target_shiny: int, tsv: int) -> int:
    if get_shiny(pid, tsv) == target_shiny:
        return pid & U32_MAX
    high = (pid & 0xFFFF) ^ (tsv & 0xFFFF) ^ (2 - target_shiny)
    return ((high << 16) | (pid & 0xFFFF)) & U32_MAX


def _force_non_shiny(pid: int, tsv: int) -> int:
    if is_shiny(pid, tsv):
        return (pid ^ 0x10000000) & U32_MAX
    return pid & U32_MAX


def _apply_non_roamer_shiny(template: StaticTemplate8, profile: Profile8, sidtid: int, pid: int) -> tuple[int, int]:
    if template.shiny == Shiny.NEVER:
        return 0, _force_non_shiny(pid, profile.tsv)

    shiny = get_shiny(pid, ((sidtid >> 16) ^ (sidtid & 0xFFFF)) & 0xFFFF)
    if shiny:
        if template.fateful:
            shiny = 2
        return shiny, _force_shiny(pid, shiny, profile.tsv)
    return 0, _force_non_shiny(pid, profile.tsv)


def _apply_roamer_shiny(profile: Profile8, sidtid: int, pid: int) -> tuple[int, int]:
    shiny = get_shiny(pid, ((sidtid >> 16) ^ (sidtid & 0xFFFF)) & 0xFFFF)
    if shiny:
        return shiny, _force_shiny(pid, shiny, profile.tsv)
    return 0, _force_non_shiny(pid, profile.tsv)


def _next_mod(rng: object, maximum: int) -> int:
    if isinstance(rng, RNGList):
        return rng.next_mod(maximum)
    return rng.next_uint(maximum)  # type: ignore[attr-defined, no-any-return]


def _next_unique_fixed_iv_indices(rng: object, fixed_count: int) -> list[int]:
    ivs = [255] * 6
    count = 0
    while count < fixed_count:
        index = _next_mod(rng, 6)
        if ivs[index] == 255:
            ivs[index] = 31
            count += 1
    return ivs


def _fill_random_ivs(rng: object, ivs: list[int]) -> tuple[int, int, int, int, int, int]:
    for index, iv in enumerate(ivs):
        if iv == 255:
            ivs[index] = _next_mod(rng, 32)
    return tuple(ivs)  # type: ignore[return-value]


def _generate_ability(rng: object, template: StaticTemplate8) -> int:
    match template.ability:
        case 0 | 1:
            return template.ability
        case 2:
            _next_mod(rng, U32_MAX + 1)
            return 2
        case _:
            return _next_mod(rng, min(template.info.ability_count, 2))


def _generate_gender(rng: object, template: StaticTemplate8, lead: Lead | int) -> int:
    ratio = template.info.gender_ratio
    match ratio:
        case value if value == GENDER_RATIO_GENDERLESS:
            return GENDER_GENDERLESS
        case value if value == GENDER_RATIO_FEMALE_ONLY:
            return GENDER_FEMALE
        case value if value == GENDER_RATIO_MALE_ONLY:
            return GENDER_MALE
        case _:
            if int(lead) in (int(Lead.CUTE_CHARM_F), int(Lead.CUTE_CHARM_M)) and _next_mod(rng, 3) > 0:
                return GENDER_MALE if int(lead) == int(Lead.CUTE_CHARM_F) else GENDER_FEMALE
            return int(_next_mod(rng, 253) + 1 < ratio)


def _generate_nature(rng: object, lead: Lead | int) -> int:
    if _is_synchronize(lead):
        return int(lead)
    return _next_mod(rng, 25)


def _generate_height_weight(rng: object) -> tuple[int, int]:
    height = _next_mod(rng, 129) + _next_mod(rng, 128)
    weight = _next_mod(rng, 129) + _next_mod(rng, 128)
    return height, weight


@dataclass(frozen=True)
class StaticGenerator8:
    """BDSP Gen 8 static encounter generator."""

    initial_advances: int
    max_advances: int
    offset: int
    lead: Lead | int
    template: StaticTemplate8
    profile: Profile8
    state_filter: StateFilter = StateFilter()

    def __post_init__(self) -> None:
        if self.initial_advances < 0:
            raise ValueError("initial_advances must be non-negative")
        if self.max_advances < 0:
            raise ValueError("max_advances must be non-negative")
        if self.offset < 0:
            raise ValueError("offset must be non-negative")

    def _generate_native(self, seed0: int, seed1: int, roamer: bool) -> list[State8]:
        """使用 C++ 原生扩展生成（高速路径），输出与 Python 完全一致。"""
        sf = self.state_filter
        template = self.template
        profile = self.profile
        args = (
            seed0, seed1,
            self.initial_advances, self.max_advances, self.offset,
            int(self.lead), roamer,
            int(template.shiny), template.fateful,
            template.iv_count, template.ability,
            template.info.gender_ratio if not roamer else 0,
            template.info.ability_count,
            profile.tid, profile.sid,
            sf.skip, sf.ability, sf.gender, sf.shiny,
            sf.height_min, sf.height_max, sf.weight_min, sf.weight_max,
            tuple(sf.iv_min), tuple(sf.iv_max),
            tuple(sf.natures), tuple(sf.hidden_powers),
        )
        try:
            results_raw = _native_generate(*args[:15], template.level, *args[15:])
        except TypeError:
            results_raw = _native_generate(*args)
        return [
            State8(
                advances=r[0], ec=r[1], sidtid=r[2], pid=r[3],
                ivs=tuple(r[4]),
                ability=r[5], gender=r[6], level=template.level,
                nature=r[8], shiny=r[9], height=r[10], weight=r[11],
            )
            for r in results_raw
        ]

    def generate(self, seed0: int | SeedPair64, seed1: int | None = None) -> list[State8]:
        if isinstance(seed0, SeedPair64):
            if seed1 is not None:
                raise ValueError("seed1 must be omitted when seed0 is a SeedPair64")
            seed0, seed1 = seed0.seeds
        if seed1 is None:
            raise ValueError("seed1 is required")
        seed0 = validate_seed64(seed0, "seed0")
        seed1 = validate_seed64(seed1, "seed1")

        if _HAS_NATIVE:
            return self._generate_native(seed0, seed1, self.template.roamer)

        if self.template.roamer:
            return self.generate_roamer(seed0, seed1)
        return self.generate_non_roamer(seed0, seed1)

    def generate_non_roamer(self, seed0: int, seed1: int) -> list[State8]:
        rng = BDSPXorshift.from_seed_pair64(SeedPair64(seed0, seed1))
        rng.advance(self.initial_advances + self.offset)
        rng_list: RNGList[int] = RNGList(rng, size=32, generate=_gen_static_ec)
        sf = self.state_filter

        states: list[State8] = []
        for count in range(self.max_advances + 1):
            rng_list.advance_state()
            ec = rng_list.next()
            sidtid = rng_list.next()
            pid = rng_list.next()
            shiny, pid = _apply_non_roamer_shiny(self.template, self.profile, sidtid, pid)

            # IV
            ivs: list[int] = [255] * 6
            fixed = 0
            while fixed < self.template.iv_count:
                idx = _next_mod(rng_list, 6)
                if ivs[idx] == 255:
                    ivs[idx] = 31
                    fixed += 1
            for iv_idx in range(6):
                if ivs[iv_idx] == 255:
                    ivs[iv_idx] = _next_mod(rng_list, 32)

            ability = _generate_ability(rng_list, self.template)
            gender = _generate_gender(rng_list, self.template, self.lead)
            nature = _generate_nature(rng_list, self.lead)
            height, weight = _generate_height_weight(rng_list)

            # 统一过滤（与 PokeFinder 一致）
            if not sf.skip:
                if sf.ability != 255 and sf.ability != ability:
                    continue
                if sf.gender != 255 and sf.gender != gender:
                    continue
                if sf.shiny != 255 and not (sf.shiny & shiny):
                    continue
                if not sf.natures[nature]:
                    continue
                if not sf.hidden_powers[hidden_power(ivs)]:
                    continue
                if height < sf.height_min or height > sf.height_max:
                    continue
                if weight < sf.weight_min or weight > sf.weight_max:
                    continue
                if not all(sf.iv_min[i] <= ivs[i] <= sf.iv_max[i] for i in range(6)):
                    continue

            states.append(State8(
                advances=self.initial_advances + count,
                ec=ec, sidtid=sidtid, pid=pid, ivs=tuple(ivs),
                ability=ability, gender=gender, level=self.template.level,
                nature=nature, shiny=shiny, height=height, weight=weight,
            ))
        return states

    def generate_roamer(self, seed0: int, seed1: int) -> list[State8]:
        import time as _time
        gender = GENDER_FEMALE if self.template.species == 488 else GENDER_GENDERLESS
        roamer = BDSPXorshift.from_seed_pair64(SeedPair64(seed0, seed1))
        roamer.advance(self.initial_advances + self.offset)
        sf = self.state_filter
        need_shiny = sf.shiny != 255

        states: list[State8] = []
        _yield_every = 50000
        for count in range(self.max_advances + 1):
            if count % _yield_every == 0:
                _time.sleep(0)
            ec = _gen_static_ec(roamer)
            rng = XoroshiroBDSP(ec)
            sidtid = rng.next_uint(U32_MAX)
            pid = rng.next_uint(U32_MAX)
            shiny, pid = _apply_roamer_shiny(self.profile, sidtid, pid)

            if need_shiny and not (sf.shiny & shiny):
                continue

            ivs = [255] * 6
            fixed = 0
            while fixed < 3:
                idx = _next_mod(rng, 6)
                if ivs[idx] == 255:
                    ivs[idx] = 31
                    fixed += 1
            for iv_idx in range(6):
                if ivs[iv_idx] == 255:
                    ivs[iv_idx] = _next_mod(rng, 32)

            ability = rng.next_uint(2)
            nature = _generate_nature(rng, self.lead)
            height, weight = _generate_height_weight(rng)

            if sf.quick_reject(shiny, ability, gender, nature, height, weight):
                continue
            if not all(sf.iv_min[i] <= ivs[i] <= sf.iv_max[i] for i in range(6)):
                continue
            if not sf.hidden_powers[hidden_power(ivs)]:
                continue

            state = State8(
                advances=self.initial_advances + count,
                ec=ec, sidtid=sidtid, pid=pid, ivs=tuple(ivs),
                ability=ability, gender=gender, level=self.template.level,
                nature=nature, shiny=shiny, height=height, weight=weight,
            )
            states.append(state)
        return states

File: src/auto_bdsp_rng/rng_core/native/generator.cpp
#include "generator.hpp"
#include "Xorshift.hpp"
#include "Xoroshiro.hpp"
#include "RNGList.hpp"

namespace {

constexpr u32 U32_MAX = 0xFFFFFFFF;
constexpr u8 GENDER_MALE = 0;
constexpr u8 GENDER_FEMALE = 1;
constexpr u8 GENDER_GENDERLESS = 2;

// PokeFinder 的 EC 生成函数: rng.next(0x80000000, 0x7FFFFFFF)
u32 gen_ec(Xorshift& rng) {
    return rng.next(0x80000000, 0x7FFFFFFF);
}

u8 get_shiny(u32 pid, u16 tsv) {
    u16 psv = ((pid >> 16) ^ (pid & 0xFFFF)) & 0xFFFF;
    if (tsv == psv) return 2;       // Square
    if ((tsv ^ psv) < 16) return 1; // Star
    return 0;
}

bool is_shiny(u32 pid, u16 tsv) {
    u16 psv = ((pid >> 16) ^ (pid & 0xFFFF)) & 0xFFFF;
    return (tsv == psv) || ((tsv ^ psv) < 16);
}

u32 force_shiny(u32 pid, u16 tsv, u8 target_shiny) {
    if (get_shiny(pid, tsv) == target_shiny) return pid & U32_MAX;
    u16 high = (pid >> 16) & 0xFFFF;
    u16 low = pid & 0xFFFF;
    high = (high ^ tsv ^ (2 - target_shiny)) & 0xFFFF;
    return ((high << 16) | low) & U32_MAX;
}

u32 force_non_shiny(u32 pid, u16 tsv) {
    if (is_shiny(pid, tsv)) {
        return (pid ^ 0x10000000) & U32_MAX;
    }
    return pid & U32_MAX;
}

bool is_synchronize(int lead) { return 0 <= lead && lead <= 24; }
bool is_cute_charm(int lead) { return lead == 25 || lead == 26; }

u8 hidden_power(const u8 ivs[6]) {
    u8 bits = 0;
    for (int i = 0; i < 6; i++) bits |= (ivs[i] & 1) << i;
    return (bits * 15) / 63;
}

bool pass_filter(const StateResult& r, const FilterParams& f) {
    if (f.skip) return true;
    if (f.ability != 255 && f.ability != r.ability) return false;
    if (f.gender != 255 && f.gender != r.gender) return false;
    if (f.shiny_mask != 255 && !(f.shiny_mask & r.shiny)) return false;
    if (!f.natures[r.nature]) return false;
    if (!f.hidden_powers[hidden_power(r.ivs)]) return false;
    if (r.height < f.height_min || r.height > f.height_max) return false;
    if (r.weight < f.weight_min || r.weight > f.weight_max) return false;
    for (int i = 0; i < 6; i++) {
        if (r.ivs[i] < f.iv_min[i] || r.ivs[i] > f.iv_max[i]) return false;
    }
    return true;
}

// 第2关快速拒绝（不含 IV/hidden power，在计算 IV 前调用）
bool quick_reject_shiny_a_g_n_h_w(u8 shiny, u8 ability, u8 gender, u8 nature,
                                   u8 height, u8 weight, const FilterParams& f) {
    if (f.skip) return false;
    if (f.ability != 255 && f.ability != ability) return true;
    if (f.gender != 255 && f.gender != gender) return true;
    if (f.shiny_mask != 255 && !(f.shiny_mask & shiny)) return true;
    if (!f.natures[nature]) return true;
    if (height < f.height_min || height > f.height_max) return true;
    if (weight < f.weight_min || weight > f.weight_max) return true;
    return false;
}

} // namespace

std::vector<StateResult> generate_non_roamer(
    u64 seed0, u64 seed1,
    u32 initial_advances, u32 max_advances, u32 offset,
    int lead, ShinyTemplate shiny_template, bool fateful,
    u8 iv_count, u8 ability_template, u8 gender_ratio, u8 ability_count,
    u16 tid, u16 sid,
    const FilterParams& filter)
{
    u16 tsv = tid ^ sid;
    bool need_shiny = filter.shiny_mask != 255;

    Xorshift rng(seed0, seed1);
    rng.jump(initial_advances + offset);
    RNGList<u32, Xorshift, 32, gen_ec> rng_list(rng);

    std::vector<StateResult> results;
    for (u32 cnt = 0; cnt <= max_advances; cnt++) {
        u32 ec = rng_list.next();
        u32 sidtid = rng_list.next();
        u32 pid = rng_list.next();

        u8 shiny;
        if (shiny_template == ShinyTemplate::Never) {
            shiny = 0;
            pid = force_non_shiny(pid, tsv);
        } else {
            u16 fake_tsv = ((sidtid >> 16) ^ (sidtid & 0xFFFF)) & 0xFFFF;
            shiny = get_shiny(pid, fake_tsv);
            if (shiny) {
                if (fateful) shiny = 2;
                pid = force_shiny(pid, tsv, shiny);
            } else {
                pid = force_non_shiny(pid, tsv);
            }
        }

        // 第1关：shiny 提前拒绝
        if (need_shiny && !(filter.shiny_mask & shiny)) {
            rng_list.advanceState();
            continue;
        }

        // IV 生成
        u8 ivs[6] = {255, 255, 255, 255, 255, 255};
        for (u8 i = 0; i < iv_count;) {
            u8 idx = rng_list.next(6);
            if (ivs[idx] == 255) { ivs[idx] = 31; i++; }
        }
        for (u8& iv : ivs) {
            if (iv == 255) iv = rng_list.next(32);
        }

        // Ability
        u8 ability;
        switch (ability_template) {
            case 0: case 1: ability = ability_template; break;
            case 2: rng_list.next(); ability = 2; break;
            default: ability = rng_list.next(ability_count > 2 ? 2 : ability_count); break;
        }

        // Gender
        u8 gender;
        switch (gender_ratio) {
            case 255: gender = GENDER_GENDERLESS; break;
            case 254: gender = GENDER_FEMALE; break;
            case 0:   gender = GENDER_MALE; break;
            default:
                if (is_cute_charm(lead) && rng_list.next(3) > 0) {
                    gender = (lead == 25) ? GENDER_MALE : GENDER_FEMALE;
                } else {
                    gender = (rng_list.next(253) + 1) < gender_ratio ? GENDER_FEMALE : GENDER_MALE;
                }
                break;
        }

        // Nature
        u8 nature;
        if (is_synchronize(lead)) {
            nature = static_cast<u8>(lead);
        } else {
            nature = rng_list.next(25);
        }

        // Height / Weight
        u8 height = rng_list.next(129) + rng_list.next(128);
        u8 weight = rng_list.next(129) + rng_list.next(128);

        // 第2关：ability/gender/nature/height/weight 提前拒绝
        if (quick_reject_shiny_a_g_n_h_w(shiny, ability, gender, nature, height, weight, filter)) {
            rng_list.advanceState();
            continue;
        }

        // 第3关：IV + hidden power（需要先获得完整 ivs）
        if (!filter.skip) {
            bool iv_pass = true;
            for (int i = 0; i < 6; i++) {
                if (ivs[i] < filter.iv_min[i] || ivs[i] > filter.iv_max[i]) {
                    iv_pass = false; break;
                }
            }
            if (!iv_pass || !filter.hidden_powers[hidden_power(ivs)]) {
                rng_list.advanceState();
                continue;
            }
        }

        StateResult r{};
        r.advances = initial_advances + cnt;
        r.ec = ec; r.sidtid = sidtid; r.pid = pid;
        for (int i = 0; i < 6; i++) r.ivs[i] = ivs[i];
        r.ability = ability; r.gender = gender; r.level = 1;
        r.nature = nature; r.shiny = shiny;
        r.height = height; r.weight = weight;
        results.push_back(r);

        rng_list.advanceState();
    }
    return results;
}

std::vector<StateResult> generate_roamer(
    u64 seed0, u64 seed1,
    u32 initial_advances, u32 max_advances, u32 offset,
    int lead, u16 tid, u16 sid, u16 species,
    const FilterParams& filter)
{
    u16 tsv = tid ^ sid;
    bool need_shiny = filter.shiny_mask != 255;
    u8 gender = (species == 488) ? GENDER_FEMALE : GENDER_GENDERLESS;

    Xorshift roamer(seed0, seed1);
    roamer.jump(initial_advances + offset);

    std::vector<StateResult> results;
    for (u32 cnt = 0; cnt <= max_advances; cnt++) {
        u32 ec = gen_ec(roamer);
        XoroshiroBDSP rng(ec);

        u32 sidtid = rng.nextUInt(U32_MAX);
        u32 pid = rng.nextUInt(U32_MAX);

        u16 fake_tsv = ((sidtid >> 16) ^ (sidtid & 0xFFFF)) & 0xFFFF;
        u8 shiny = get_shiny(pid, fake_tsv);
        if (shiny) {
            pid = force_shiny(pid, tsv, shiny);
        } else {
            pid = force_non_shiny(pid, tsv);
        }

        if (need_shiny && !(filter.shiny_mask & shiny)) continue;

        // 3 fixed 31 IVs
        u8 ivs[6] = {255, 255, 255, 255, 255, 255};
        for (int i = 0; i < 3;) {
            u8 idx = rng.nextUInt(6);
            if (ivs[idx] == 255) { ivs[idx] = 31; i++; }
        }
        for (u8& iv : ivs) {
            if (iv == 255) iv = rng.nextUInt(32);
        }

        u8 ability = rng.nextUInt(2);

        u8 nature;
        if (is_synchronize(lead)) {
            nature = static_cast<u8>(lead);
        } else {
            nature = rng.nextUInt(25);
        }

        u8 height = rng.nextUInt(129) + rng.nextUInt(128);
        u8 weight = rng.nextUInt(129) + rng.nextUInt(128);

        if (quick_reject_shiny_a_g_n_h_w(shiny, ability, gender, nature, height, weight, filter)) continue;
        if (!filter.skip) {
            bool iv_pass = true;
            for (int i = 0; i < 6; i++) {
                if (ivs[i] < filter.iv_min[i] || ivs[i] > filter.iv_max[i]) {
                    iv_pass = false; break;
                }
            }
            if (!iv_pass || !filter.hidden_powers[hidden_power(ivs)]) continue;
        }

        StateResult r{};
        r.advances = initial_advances + cnt;
        r.ec = ec; r.sidtid = sidtid; r.pid = pid;
        for (int i = 0; i < 6; i++) r.ivs[i] = ivs[i];
        r.ability = ability; r.gender = gender; r.level = 1;
        r.nature = nature; r.shiny = shiny;
        r.height = height; r.weight = weight;
        results.push_back(r);
    }
    return results;
}
